File: R/parse.R
parse_distance_text <- function(distance_string) {
  #browser()
  distrance_string_clean <- distance_string |>
    stringr::str_remove_all("\\s") |>
    stringr::str_to_lower()

  number = NULL
  unit = NULL

  number = stringr::str_extract(distrance_string_clean, "\\d*") |>
    as.numeric()
  unit = stringr::str_extract(distrance_string_clean, "[a-z]+")

  # convert meters to km
  if (unit == "m") {
    unit = "km"
    number = number / 1000
  } else if (unit == "half") {
    unit = "km"
    number = 21.1
  } else if (unit == "marathon") {
    unit = "mi"
    number = 26.2
  }

  distance_unit <- list(
    distance = number,
    unit = unit
  )
  class(distance_unit) <- "distance_unit"
  return(distance_unit)
}

expected <- list(distance = 0.4, unit = "km")
class(expected) <- "distance_unit"
testthat::expect_equal(parse_distance_text("   400 m   "), expected)

expected <- list(distance = 10, unit = "km")
class(expected) <- "distance_unit"
testthat::expect_equal(parse_distance_text("   10 km   "), expected)

File: pyrunpacer/src/pyrunpacer/parse.py
import re


def parse_distance_text(distance_string):
    """Parses a string into its distance and unit (in kilometers or miles)

    Returns a dictionary with keys: "distance", and "unit", e.g.,
    {"distance": 42.2, "unit": "km"}
    """
    distance_string_clean = re.sub(
        pattern=r"\s+", repl="", string=distance_string
    ).lower()

    try:
        number = float(re.search(r"\d+", distance_string_clean).group(0))
    except AttributeError:
        # when we pass in only a string, e.g., "marathon", or "half"
        number = None

    # the unit is just the string, we convert to numeric distances below
    unit = str(re.search(r"[a-zA-Z]+", distance_string_clean).group(0))

    # get the distance and units in km or mi
    if unit == "km":
        unit = "km"
        number = number
    elif unit == "m":
        unit = "km"
        number = number / 1_000
    elif unit == "mi":
        unit = "mi"
        number = number
    elif unit == "half":
        unit = "km"
        number = 21.1
    elif unit == "marathon":
        unit = "km"
        number = 42.2
    else:
        raise ValueError(
            f'Unknown unit passed. Got unit: "{unit}" and number: "{number}"'
        )

    distance_unit = {"distance": number, "unit": unit}

    return distance_unit


assert parse_distance_text("marathon") == {"distance": 42.2, "unit": "km"}
assert parse_distance_text("half") == {"distance": 21.1, "unit": "km"}
assert parse_distance_text("400m") == {"distance": 0.4, "unit": "km"}
assert parse_distance_text("5 km") == {"distance": 5, "unit": "km"}
assert parse_distance_text("10 mi") == {"distance": 10, "unit": "mi"}
